src/rules/maxNestingDepthRule.ts

```diff
@@ -0,0 +1,142 @@
+            It is computed by determining the nesting levels of the control statements within the method.
+            A starting value of 0 is assigned, then this value is
+            incremented each time a new block is entered, and decremented
+            each time a block is left.
+            The following statements contribute to nesting depth:
+            * \`if\` statements
+            * \`switch\` statements
+            * \`for\`, \`for in\` and \`for of\` loops
+            * \`while\` and \`do while\` loops
+            * \`try\` \`catch\` and \`finally\` blocks`,
+        rationale: Lint.Utils.dedent`
+            Maximum nesting depth is a code metric which indicates the level of imbrication in a
+            function. This metric quickly reveals the methods whose code is made difficult to read
+            by an excessive nesting of the control statements.
+
+            It's better to have smaller, single-purpose functions with self-documenting names.`,
+        optionsDescription: Lint.Utils.dedent`
+            An optional upper limit for maximum nesting depth can be specified. If no limit option
+            is provided a default value of ${Rule.DEFAULT_THRESHOLD} will be used.`,
+        options: {
+            type: "number",
+            minimum: Rule.MINIMUM_THRESHOLD,
+        },
+        optionExamples: [true, [true, 5]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(expected: number, actual: number, name?: string): string {
+        return (
+            `The function${name === "" ? "" : ` ${name}`} has a nesting depth of at least ` +
+            `${actual} which is higher than the threshold of ${expected}`
+        );
+    }
+
+    public isEnabled(): boolean {
+        // Disable the rule if the option is provided but non-numeric or less than the minimum.
+        const isThresholdValid =
+            typeof this.threshold === "number" && this.threshold >= Rule.MINIMUM_THRESHOLD;
+        return super.isEnabled() && isThresholdValid;
+    }
+
+    public get threshold(): number {
+        if (this.ruleArguments[0] !== undefined) {
+            return this.ruleArguments[0] as number;
+        }
+        return Rule.DEFAULT_THRESHOLD;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk, { threshold: this.threshold });
+    }
+}
+
+function getFunctionName(node: ts.FunctionLikeDeclaration): string {
+    return node.name !== undefined && isIdentifier(node.name) ? node.name.text : "";
+}
+
+function walk(ctx: Lint.WalkContext<{ threshold: number }>): void {
+    const {
+        options: { threshold },
+    } = ctx;
+    const nestingDepthStack: number[] = [];
+    const functionNameStack: string[] = [];
+    const cbNode = (node: ts.Node): void => {
+        if (isFunctionScopeBoundary(node)) {
```

Comment:
> `isFunctionScopeBoundary`

I believe you can use `isFunctionWithBody` from `tsutils` so that you don't need the `as` cast a few lines down.

Reply:
`isFunctionWithBody` is indeed more appropriate since the needed tests are narrowed, but as it is also returning a boolean I still need the `as` cast.

Reply:
Oh! I missed the declaration, you were right!